pkg/ee/constraint-controller/util.go

```diff
@@ -0,0 +1,178 @@
+
+   END OF TERMS AND CONDITIONS
+*/
+
+package seedconstraintsynchronizer
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	kubermaticv1 "k8c.io/kubermatic/v2/pkg/crd/kubermatic/v1"
+	"k8c.io/kubermatic/v2/pkg/provider"
+
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/sets"
+	ctrlruntimeclient "sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	Key = "default"
+)
+
+type clusterNamesList []string
+
+func getClusterNameFromNamespace(clusterNamespace string) string {
+	var clusterName string
+	temp := strings.Split(clusterNamespace, "-")
+	if len(temp) == 2 {
+		clusterName = temp[1]
+	}
+	return clusterName
+}
+
+func getClusterNamesForExistingConstraint(ctx context.Context, client ctrlruntimeclient.Client, constraint *kubermaticv1.Constraint) (clusterNamesList, error) {
+
+	var existingClusterNames []string
+
+	labelSelector, err := labels.Parse(fmt.Sprintf("%s=%s", Key, constraint.Name))
+	if err != nil {
+		return nil, err
+	}
+
+	constraintList := &kubermaticv1.ConstraintList{}
+
+	err = client.List(ctx, constraintList, &ctrlruntimeclient.ListOptions{LabelSelector: labelSelector})
+	if err != nil {
+		return nil, err
+	}
+
+	for _, defaultConstraint := range constraintList.Items {
+		clusterName := getClusterNameFromNamespace(defaultConstraint.Namespace)
+		existingClusterNames = append(existingClusterNames, clusterName)
+	}
+	return existingClusterNames, nil
+}
+
+func getClusterList(ctx context.Context, client ctrlruntimeclient.Client, clusterNames clusterNamesList) []kubermaticv1.Cluster {
+	var existingClusterList []kubermaticv1.Cluster
+	for _, clusterName := range clusterNames {
+		cluster := &kubermaticv1.Cluster{}
+		err := client.Get(ctx, types.NamespacedName{Name: clusterName}, cluster)
+		if err != nil {
+			return nil
+		}
+		existingClusterList = append(existingClusterList, *cluster)
+	}
+
+	return existingClusterList
+}
+
+func (s clusterNamesList) contains(searchterm string) bool {
+	for _, value := range s {
+		if value == searchterm {
+			return true
+		}
+	}
+	return false
+}
+
+func (s clusterNamesList) difference(s2 clusterNamesList) []string {
+	var result []string
+	for _, value := range s {
+		if !s2.contains(value) {
+			result = append(result, value)
+		}
+	}
+	return result
+}
+
+// GetClustersForConstraint gets clusters for the constraints by using the constraints selector to filter out unselected clusters
+func GetClustersForConstraint(ctx context.Context, client ctrlruntimeclient.Client,
+	constraint *kubermaticv1.Constraint, workerNamesLabelSelector labels.Selector) ([]kubermaticv1.Cluster, []kubermaticv1.Cluster, error) {
+	var desiredClusterNames clusterNamesList
+
+	desiredList, err := getDesiredClusterListForConstraint(ctx, client, constraint, workerNamesLabelSelector)
+	if err != nil {
+		return nil, nil, fmt.Errorf("failed listing clusters: %w", err)
+	}
+
+	desiredClusterList := desiredList.Items
+
```

Comment:
I think a simpler solution here would be to get all clusters (simply with client like in the CE code) and then just do a diff of the desiredClusterList and allClustersList .

Reply:
But that list will contain clusters where constraint is not present. suppose we have 15 clusters and constraints are present on only 3 desired clusters then and 2 as unwanted. then instead of running the loop for 2 clusters (for deletion) we will be running the loop for all the 12 clusters unnecessarily and probably the logs will show that we are trying to delete the resource which is not present

Reply:
Thats true, but I am not sure if that bothers us. We can eventually change the logs to log after a succesfull deletion, and just skip error log when not found.

But I think its a simpler and less bug-prone solution. @irozzo-1A WDYT?

Reply:
I think that we can start simply as @lsviben is suggesting and favor readability and robustness. I'm not concerned about the logs, but definitely, we can save some API calls. I would rather do that in `cleanupConstraint` by providing the list of current constraints so that we are not calling Delete unconditionally (BTW this is an optimization and can be done in a follow-up PR).

Note that if we build on top of what I suggested before, you could simply implement a filtering function, that could go through the list of all clusters given in input and avoid additional API calls.